lib/RestructureCFG/ScopeGraphUtils.cpp

```diff
@@ -0,0 +1,187 @@
+/// \file ScopeGraphUtils.cpp
+/// Helpers for the `ScopeGraph` building
+///
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/LLVMContext.h"
+
+#include "revng/RestructureCFG/ScopeGraphUtils.h"
+#include "revng/Support/IRHelpers.h"
+
+using namespace llvm;
+
+// Helper function which set the attributes for the created function
+// prototypes
+static void setFunctionAttributes(Function *F, FunctionTags::Tag &Tag) {
+  F->setLinkage(GlobalValue::ExternalLinkage);
+  F->addFnAttr(Attribute::OptimizeNone);
+  F->addFnAttr(Attribute::NoInline);
+  F->addFnAttr(Attribute::NoMerge);
+  F->addFnAttr(Attribute::NoUnwind);
+  F->addFnAttr(Attribute::WillReturn);
+  F->setMemoryEffects(MemoryEffects::inaccessibleMemOnly());
+
+  // Add the tag to the `Function`
+  Tag.addTo(F);
+}
+
+void ScopeGraphBuilder::initScopeCloserFunction(Function *F) {
+  LLVMContext &C = getContext(F);
+  Module *M = getModule(F);
+  FunctionTags::Tag &Tag = FunctionTags::ScopeCloserMarker;
+  ScopeCloserFunction = getUniqueFunctionWithTag(Tag, M);
+
+  // Create the `ScopeCloserMarker` function if it doesn't exists
+  if (not ScopeCloserFunction) {
+    auto *FT = FunctionType::get(Type::getVoidTy(C), {}, false);
+    ScopeCloserFunction = cast<Function>(M->getOrInsertFunction(Tag.name(), FT)
+                                           .getCallee());
+    setFunctionAttributes(ScopeCloserFunction, Tag);
+  }
+}
```

Comment:
Like I already said in another comment. This can be a `static` free function.

```c++
Function *getOrCreateScopeCloserFunction(Module *M) {
  FunctionTags::Tag &ScopeCloser = FunctionTags::ScopeCloserMarker;
  Function *Result = getUniqueFunctionWithTag(ScopeCloser, M);
  if (not Result) {
    auto *FT = FunctionType::get(Type::getVoidTy(M->getContext()), {}, false);
    llvm::FunctionCallee F = M->getOrInsertFunction(ScopeCloser.name(), FT);
    Result = cast<Function>(F.getCallee());
    setFunctionAttributes(Result, ScopeCloser);
  }
  revng_assert(Result != nullptr);
  return Result;
}
```

And then the only caller just does:
```c++
ScopeCloserFunction = getOrCreateScopeCloserFunction(F->getParent());
```

You can even go and specify the `ScopeGraphBuilder` constructor as:
```c++
ScopeGraphBuilder::ScopeGraphBuilder(Function *F) : 
    ScopeCloserFunction(getOrCreateScopeCloserFunction(F)),
    GotoBlockFunction(getOrCreateGotoBlockFunction(F)) {
}
```

Reply:
When this is done we can merge this